collect/collect.go

```diff
@@ -228,6 +245,77 @@ func (i *InMemCollector) checkAlloc() {
 	runtime.GC()
 }
 
+func (i *InMemCollector) newCheckAlloc() {
+	inMemConfig, err := i.Config.GetInMemCollectorCacheCapacity()
+
+	var mem runtime.MemStats
+	runtime.ReadMemStats(&mem)
+	i.Metrics.Gauge("memory_heap_allocation", int64(mem.Alloc))
+	if err != nil || inMemConfig.MaxAlloc == 0 || mem.Alloc < inMemConfig.MaxAlloc {
+		return
+	}
+
+	// Figure out what fraction of the total cache we should remove. We'd like it to be
+	// enough to get us below the max capacity, but not TOO much below.
+	// Because our impact numbers are only the data size, reducing by enough to reach
+	// max alloc will actually do more than that.
+	totalToRemove := mem.Alloc - inMemConfig.MaxAlloc
+
+	// the size of the cache exceeds the user's intended allocation, so we're going to
+	// remove the traces from the cache that have had the most impact on allocation.
+	// To do this, we sort the traces by their CacheImpact value and then remove traces
+	// until the total size is less than 90% of the intended max.
+
+	existingCache, ok := i.cache.(*cache.DefaultInMemCache)
+	if !ok {
+		i.Logger.Error().WithField("alloc", mem.Alloc).Logf(
+			"total allocation exceeds limit, but unable to control cache",
+		)
+		return
+	}
+	allTraces := existingCache.GetAll()
+	timeout, err := i.Config.GetTraceTimeout()
+	if err != nil {
+		timeout = 60 * time.Second
+	} // Sort traces by CacheImpact, heaviest first
+	sort.Slice(allTraces, func(i, j int) bool {
+		return allTraces[i].CacheImpact(timeout) > allTraces[j].CacheImpact(timeout)
```

Comment:
Since `CacheImpact` may be called multiple times for some traces, it _may_ be worth memoizing. It would take a benchmark to be more sure.

Reply:
Good catch! I did a little abstract benchmark of Sort() and it calls the comparator `O(Nlog(N))` times! So even without testing here, it seemed well worth memoizing this.